src/roads.cpp

```diff
@@ -8,13 +8,411 @@
 #include <osg/MatrixTransform>
 #include <osg/ShapeDrawable>
 
+#include <osg/Geometry>
+#include <osg/Geode>
+#include <osg/Texture2D>
+#include <osg/Program>
+#include <osg/Shader>
+#include <osg/Material>
+#include <osgSim/ShapeAttribute>
+
 #include <iostream>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <cmath>
 
 #include "common.h"
 
 using namespace osg;
 
-osg::Node* process_roads(osg::Matrixd& ltw, const std::string & file_path)
+static const char* vertSource = R"(
+    #version 120
+    attribute vec3 a_tangent; 
+    varying vec2 v_texCoord;
+    varying vec3 v_lightDir;
+    varying vec3 v_viewDir;
+
+    void main() {
+        v_texCoord = gl_MultiTexCoord0.xy;
+        vec3 ecPos = vec3(gl_ModelViewMatrix * gl_Vertex);
+        
+        vec3 n = normalize(gl_NormalMatrix * gl_Normal);
+        vec3 t = normalize(gl_NormalMatrix * a_tangent);
+        vec3 b = cross(n, t);
+        mat3 tbn = mat3(t, b, n);
+
+        vec3 lightPos = gl_LightSource[0].position.xyz;
+        v_lightDir = (lightPos - ecPos) * tbn;
+        v_viewDir = -ecPos * tbn;
+
+        gl_Position = gl_ModelViewProjectionMatrix * gl_Vertex;
+    }
+)";
+
+static const char* fragSource = R"(
+    #version 120
+    uniform sampler2D diffuseMap;
+    uniform sampler2D normalMap;
+    varying vec2 v_texCoord;
+    varying vec3 v_lightDir;
+    varying vec3 v_viewDir;
+
+    void main() {
+        vec4 texColor = texture2D(diffuseMap, v_texCoord);
+        vec3 N = normalize(texture2D(normalMap, v_texCoord).rgb * 2.0 - 1.0);
+        vec3 L = normalize(v_lightDir);
+        vec3 V = normalize(v_viewDir);
+        
+        float NdotL = max(dot(N, L), 0.0);
+        vec3 H = normalize(L + V);
+        float NdotH = max(dot(N, H), 0.0);
+        
+        vec3 ambient = gl_LightSource[0].ambient.rgb * texColor.rgb * 0.3;
+        vec3 diffuse = gl_LightSource[0].diffuse.rgb * texColor.rgb * NdotL;
+        vec3 specular = gl_LightSource[0].specular.rgb * pow(NdotH, 32.0) * 0.2;
+
+        gl_FragColor = vec4(ambient + diffuse + specular, texColor.a);
+    }
+)";
+
+inline std::string trim(const std::string& str)
+{
+    const char* ws = " \t\n\r\f\v";
+    size_t start = str.find_first_not_of(ws);
+    if (start == std::string::npos) return "";
+    size_t end = str.find_last_not_of(ws);
+    return str.substr(start, end - start + 1);
+}
+
+osg::StateSet* createTextureStateSet(osg::Program* program, const std::string& diffPath, const std::string& normPath)
+{
+    osg::StateSet* ss = new osg::StateSet();
+    ss->setAttributeAndModes(program, osg::StateAttribute::ON);
+    ss->addUniform(new osg::Uniform("diffuseMap", 0));
+    ss->addUniform(new osg::Uniform("normalMap", 1));
+
```

Comment:
```suggestion

    ss->setAttributeAndModes(new osg::Depth(osg::Depth::LESS, 0, 1, false));
    ss->setRenderBinDetails(order, "RenderBin");
    ss->setNestRenderBins(false);
```

disable depth writes (zmask) and apply order of drawing.